### ISSUE TITLE
pocelain does not merge updates into working trees

### ISSUE THREAD

**Issue by Salamafet** [jelmer/dulwich#452]
When a file is removed from repository, the dulwich.pull command does not remove the file from local storage.

When I add a file in repository, the file is correctly downloaded.

I have try with admin right but doesn't work.

**Comment by jelmer** [comment]
On Mon, Sep 05, 2016 at 06:49:26AM -0700, Stéphane wrote:
> When a file is removed from repository, the dulwich.pull command does not remove the file from local storage.
> 
> When I add a file in repository, the file is correctly downloaded.
> 
> I have try with admin right but doesn't work.

Dulwich currently doesn't do any merging, including merges that affect the working tree.

**Comment by sheluchin** [comment]
> Dulwich currently doesn't do any merging, including merges that _affect_ the working tree.
@jelmer Is this still the case?

**Comment by jelmer** [comment]
Yes, this limitation still exists. I'm keen to get this fixed before 0.20.

On 23 January 2019 15:51:13 GMT, Alex Sheluchin <notifications@github.com> wrote:
>> Dulwich currently doesn't do any merging, including merges that
>_affect_ the working tree.
>@jelmer Is this still the case?
>
>-- 
>You are receiving this because you were mentioned.
>Reply to this email directly or view it on GitHub:
>https://github.com/dulwich/dulwich/issues/452#issuecomment-456853995

-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

**Comment by laneb** [comment]
"Merge" here refers to the update to WD+index after a clean merge, checkout, etc., correct? I.e. the WD+index reflected tree T<sub>0</sub> and we want to update WD+index to reflect T<sub>1</sub>. I think this operation can be implemented naively as

1. Assert index matches T<sub>0</sub> 
2. Assert T<sub>1</sub> contains no blob at the same path as an untracked file in WD
3. Remove all files tracked in T<sub>0</sub> from WD
4. Write all files tracked in T<sub>1</sub> to WD
5. Update index ti

@jelmer Does that seem correct? Not sure what the gotchas here are. 

Of course if there

**Comment by jelmer** [comment]
No, it refers to a three-way merge. You're describing a fast-forward operation, which would also be useful to implement but is different.

**Comment by laneb** [comment]
Oh duh 🤦‍♂️Yes I was only considering the fast-forward scenario. Makes sense that WD cannot be updated until after 3-way merge. Once that's complete and the ref is updated to point to the merge commit, the operation I can be applied.

**Comment by jelmer** [comment]
The merge has to be applied in the working directory, rather than as a commit. This is because there can be conflicts.

**Comment by kevinhendricks** [comment]
Is there any update on this?  Would it help in any way to take the Sigil dulwich repo with its preliminary merge code from about 2 years ago and update it to current dulwich?  Is there anything I can do to help get basic merge working the way you want?
